src/jdk.naming.dns/share/classes/com/sun/jndi/dns/DnsClient.java

```diff
@@ -471,7 +472,9 @@ private byte[] doUdpQuery(Packet pkt, InetAddress server,
                             return cachedMsg;
                         }
                     }
-                    timeoutLeft = pktTimeout - ((int) (end - start));
+                    long elapsedMillis = TimeUnit.NANOSECONDS.toMillis(end - start);
+                    // Setting the Math.clamp min to 1 ensures that the timeout is decreased
+                    timeoutLeft = timeoutLeft - Math.clamp(elapsedMillis, 1, Integer.MAX_VALUE);
```

Comment:
I think I'd prefer to calculate the remaining timeout based on the total elapsed time in this loop, as opposed to the time spent in blockingReceive.

Reply:
I have been wondering why the timeout was measured in this way. My explanation was that the original author of the API (duke? ;-) ) wanted to measure "actual timeout" (time actually spent waiting for a packet) as opposed to perceived timeout (time the caller spent waiting). Rationale might have been to exclude potential time spent waiting for a lock - for instance. That said - I'm not opposed to extend the scope of the timeout to make it closer to the perceived timeout, which is what the test expect. 
If we do this maybe we could move `start` out of the `do` block  - which would make the computation of `timeoutLeft` simpler (we wouldn't need the `-=`).

Reply:
I think that "the perceived timeout" described by you above is closer to the following description of the timeout environment property:
```The provider submits a query to a DNS server and waits for a response to arrive within a timeout period (1 second by default)```
And the query is submitted to a DNS server right before entrering the while loop.
I will apply the suggested simplification of the timeout left now, thanks.

Reply:
The code now measures "the perceived timeout": [4d33d05](https://github.com/openjdk/jdk/pull/20892/commits/4d33d05b111f38a418dac57587236465f2359f8c)
The timeout JNDI/DNS tests seem to be stable and a bit more accurate across multiple runs with the latest changes.